File: app/src/main/java/marcosmiranda/seniasnicas/Diccionario.java
package marcosmiranda.seniasnicas;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import androidx.appcompat.app.AppCompatActivity;

import com.marcosmiranda.seniasnicas.R;

public class Diccionario extends AppCompatActivity {
    public static final String EXTRA_CATEGORY = String.valueOf(R.string.extra_category);
    //public static String category;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_diccionario);
    }

    public void anatomia(View view) {
        Intent intent = new Intent(this, AnatomiaWordList.class);
        //category = String.valueOf(R.string.categoria_anatomia).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void animales(View view) {
        Intent intent = new Intent(this, AnimalesWordList.class);
        //category = String.valueOf(R.string.categoria_animales).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void colores(View view) {
        Intent intent = new Intent(this, ColoresWordList.class);
        //category = String.valueOf(R.string.categoria_colores).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void comida(View view) {
        Intent intent = new Intent(this, ComidaWordList.class);
        //category = String.valueOf(R.string.categoria_comida).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void direcciones(View view) {
        Intent intent = new Intent(this, DireccionesWordList.class);
        //category = String.valueOf(R.string.categoria_direcciones).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void emergencias(View view) {
        Intent intent = new Intent(this, EmergenciasWordList.class);
        //category = String.valueOf(R.string.categoria_emergencias).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void familia(View view) {
        Intent intent = new Intent(this, FamiliaWordList.class);
        //category = String.valueOf(R.string.categoria_familia).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void geografia(View view) {
        Intent intent = new Intent(this, GeografiaWordList.class);
        //category = String.valueOf(R.string.categoria_geografia).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void numeros(View view) {
        Intent intent = new Intent(this, NumerosWordList.class);
        //category = String.valueOf(R.string.categoria_numeros).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void ropa(View view) {
        Intent intent = new Intent(this, RopaWordList.class);
        //category = String.valueOf(R.string.categoria_ropa).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void saludos(View view) {
        Intent intent = new Intent(this, SaludosWordList.class);
        //category = String.valueOf(R.string.categoria_saludos).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void tiempo(View view) {
        Intent intent = new Intent(this, TiempoWordList.class);
        //category = String.valueOf(R.string.categoria_tiempo).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void trabajos(View view) {
        Intent intent = new Intent(this, TrabajosWordList.class);
        //category = String.valueOf(R.string.categoria_trabajos).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void transporte(View view) {
        Intent intent = new Intent(this, TransporteWordList.class);
        //category = String.valueOf(R.string.categoria_transporte).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }

    public void verbos(View view) {
        Intent intent = new Intent(this, VerbosWordList.class);
        //category = String.valueOf(R.string.categoria_verbos).toUpperCase();
        //intent.putExtra(EXTRA_CATEGORY, category);
        startActivity(intent);
    }
}

File: app/src/main/java/com/marcosmiranda/seniasnicas/Diccionario.kt
package com.marcosmiranda.seniasnicas

import android.app.Activity
import android.content.Intent
import android.content.res.ColorStateList
import android.content.res.Resources.Theme
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.View
import android.widget.Button
import android.widget.TableLayout
import android.widget.TableRow
import androidx.core.view.marginBottom
import androidx.room.Room

class Diccionario : Activity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_diccionario)

        val tlCategorias = findViewById<TableLayout>(R.id.activity_diccionario_tl_categorias)

        val btnLayoutParams = TableRow.LayoutParams(
            450,
            TableRow.LayoutParams.WRAP_CONTENT
        )
        btnLayoutParams.setMargins(32, 32, 32, 32)
        val trLayoutParams = TableRow.LayoutParams()
        trLayoutParams.setMargins(0, 32, 0, 32)

        val db = Room.databaseBuilder(applicationContext, Database::class.java, "SeniasNicas")
            .createFromAsset("databases/SeniasNicas.db")
            .allowMainThreadQueries()
            .fallbackToDestructiveMigration()
            .build()

        val categoriaDao = db.categoriaDao()
        val categorias = categoriaDao.getAll()
        // Log.e("categorias", categorias.toString())

        val categoriasPerRow = 2
        var c = 0
        var row = TableRow(applicationContext)
        categorias.forEach { ctg ->
            // Log.e("categoria", ctg.toString())
            val ctgId = ctg.id
            val ctgNombre = ctg.nombre

            val ctgHasPalabras = db.palabraDao().getByCategoria(ctgId).isNotEmpty()
            if (!ctgHasPalabras) return@forEach

            val btnIconId: Int
            val btnColorId: Int
            when (ctgId) {
                1 -> {
                    // Anatomía
                    btnIconId = R.drawable.person_icon
                    btnColorId = R.drawable.tan_button
                }
                2 -> {
                    // Animales
                    btnIconId = R.drawable.paw_icon
                    btnColorId = R.drawable.brown_button
                }
                3 -> {
                    // Colores
                    btnIconId = R.drawable.palette_icon
                    btnColorId = R.drawable.rgb_button
                }
                4 -> {
                    // Comida
                    btnIconId = R.drawable.utensils_icon
                    btnColorId = R.drawable.olive_button
                }
                // 5 -> {
                //     // Cristianismo
                //     btnIconId = R.drawable.cross_icon
                //     btnColorId = R.drawable.white_button
                // }
                6 -> {
                    // Deportes
                    btnIconId = R.drawable.soccer_icon
                    btnColorId = R.drawable.tan_button
                }
                7 -> {
                    // Direcciones
                    btnIconId = R.drawable.compass_icon
                    btnColorId = R.drawable.white_button
                }
                8 -> {
                    // Emergencias
                    btnIconId = R.drawable.truck_medical_icon
                    btnColorId = R.drawable.red_button
                }
                9 -> {
                    // Familia
                    btnIconId = R.drawable.people_roof_icon
                    btnColorId = R.drawable.pink_button
                }
                10 -> {
                    // Geografía
                    btnIconId = R.drawable.earth_icon
                    btnColorId = R.drawable.green_button
                }
                // 11 -> {
                //     // Herramientas
                //     btnIconId = R.drawable.sign_language_icon
                //     btnColorId = R.drawable.tan_button
                // }
                // 12 -> {
                //     // Lugares
                //     btnIconId = R.drawable.sign_language_icon
                //     btnColorId = R.drawable.tan_button
                // }
                13 -> {
                    // Números
                    btnIconId = R.drawable.numbers_icon
                    btnColorId = R.drawable.purple_button
                }
                14 -> {
                    // Ropa
                    btnIconId = R.drawable.shirt_icon
                    btnColorId = R.drawable.magenta_button
                }
                15 -> {
                    // Saludos
                    btnIconId = R.drawable.handshake_icon
                    btnColorId = R.drawable.white_button
                }
                16 -> {
                    // Tiempo
                    btnIconId = R.drawable.clock_icon
                    btnColorId = R.drawable.cyan_teal_button
                }
                17 -> {
                    // Trabajos
                    btnIconId = R.drawable.graduation_cap_icon
                    btnColorId = R.drawable.blue_navy_button
                }
                18 -> {
                    // Transporte
                    btnIconId = R.drawable.car_side_icon
                    btnColorId = R.drawable.gray_button
                }
                19 -> {
                    // Verbos
                    btnIconId = R.drawable.bed_icon
                    btnColorId = R.drawable.tan_button
                }
                else -> {
                    btnIconId = 0
                    btnColorId = 0
                }
            }

            val btn = Button(applicationContext)
            btn.text = ctgNombre
            // btn.textSize = R.dimen.btn_text_size.toFloat()
            btn.setTextColor(Color.WHITE)
            btn.setShadowLayer(5f, 0f, 0f, Color.BLACK)
            btn.layoutParams = btnLayoutParams
            btn.setCompoundDrawablesWithIntrinsicBounds(btnIconId, 0, 0, 0)
            btn.setBackgroundResource(btnColorId)
            btn.gravity = Gravity.CENTER
            // btn.setPadding(8, 0, 8, 0)
            btn.setOnClickListener {
                val intent = Intent(this, CategoriaPalabras::class.java)
                intent.putExtra("categoria", ctgId)
                intent.putExtra("icon", btnIconId)
                startActivity(intent)
            }
            row.addView(btn)
            c++

            if (c == categoriasPerRow || ctg == categorias.last()) {
                row.gravity = Gravity.CENTER
                // row.layoutParams = trLayoutParams
                tlCategorias.addView(row)
                row = TableRow(applicationContext)
                c = 0
            }
        }
    }
}